File: samples/java/DeliveryDispatch/Shared/src/main/java/systems/zlink/samples/deliverydispatch/shared/contracts/Messages.java
package systems.zlink.samples.deliverydispatch.shared.contracts;

import systems.zlink.framework.actors.ActorRefSnapshot;
import systems.zlink.framework.handlers.ZLinkPacket;

public final class Messages {
    private Messages() {}

    public enum DeliveryStatus {
        Created,
        Assigned,
        Accepted,
        Reassigned,
        PickedUp,
        Delivered,
        Failed
    }

    // --8<-- [start:doc-explicit-packet-name]
    @ZLinkPacket("CreateDeliveryReq")
    public record CreateDeliveryReq(
            String deliveryId, String customerId, String pickupAddress, String dropoffAddress) {}

    // --8<-- [end:doc-explicit-packet-name]

    public record CreateDeliveryRes(String deliveryId) {}

    @ZLinkPacket("SubscribeDeliveryReq")
    public record SubscribeDeliveryReq(String deliveryId) {}

    public record SubscribeDeliveryRes(String deliveryId) {}

    public record DeliveryStatusNotify(
            String deliveryId, DeliveryStatus status, String courierId, String occurredAt) {}

    @ZLinkPacket("AssignDeliveryMsg")
    public record AssignDeliveryMsg(
            String deliveryId, String customerId, String pickupAddress, String dropoffAddress) {}

    @ZLinkPacket("BindCourierSessionReq")
    public record BindCourierSessionReq(
            String courierId, ActorRefSnapshot actor, String sessionRoute) {
        public BindCourierSessionReq(String courierId) {
            this(courierId, null, null);
        }
    }

    public record BindCourierSessionRes(
            String courierId, ActorRefSnapshot actor, String sessionRoute) {}

    @ZLinkPacket("BindCourierReq")
    public record BindCourierReq(String courierId, String sessionRoute) {}

    public record BindCourierRes(String courierId, ActorRefSnapshot actor, String sessionRoute) {}

    @ZLinkPacket("FindCourierActorReq")
    public record FindCourierActorReq(String courierId) {}

    public record FindCourierActorRes(String courierId, ActorRefSnapshot actor) {}

    @ZLinkPacket("EnsureCourierActorReq")
    public record EnsureCourierActorReq(String courierId) {}

    public record EnsureCourierActorRes(String courierId, ActorRefSnapshot actor) {}

    /**
     * The offer, and the courier's answer to it. Both are one-way (common sample spec section 7.4):
     * a courier looks at a screen and presses a button, and tying that time to a request's reply
     * would hold the spot's serial queue open for as long as the courier thinks. The decision was
     * always going to arrive as a separate inbound message, because the server can only push to a
     * client, never request of it. {@code attempt} is what makes the pairing safe: a decision that
     * names another attempt arrived too late and is dropped.
     */
    @ZLinkPacket("OfferDeliveryMsg")
    public record OfferDeliveryMsg(
            String courierId,
            String deliveryId,
            int attempt,
            String pickupAddress,
            String dropoffAddress) {}

    @ZLinkPacket("OfferDeliveryResultMsg")
    public record OfferDeliveryResultMsg(
            String deliveryId, String courierId, int attempt, boolean accepted, String reason) {}

    public record OfferDeliveryNotify(
            String courierId, String deliveryId, String pickupAddress, String dropoffAddress) {}

    @ZLinkPacket("CourierDecisionMsg")
    public record CourierDecisionMsg(
            String deliveryId, String courierId, boolean accepted, String reason) {}

    @ZLinkPacket("ReassignDeliveryMsg")
    public record ReassignDeliveryMsg(
            String deliveryId, String previousCourierId, String nextCourierId, String reason) {}

    @ZLinkPacket("DeliveryStatusChangedReq")
    public record DeliveryStatusChangedReq(
            String deliveryId, DeliveryStatus status, String courierId, String occurredAt) {}

    @ZLinkPacket("DeliveryStatusUpdatedMsg")
    public record DeliveryStatusUpdatedMsg(
            String deliveryId,
            String customerId,
            DeliveryStatus status,
            String courierId,
            String occurredAt) {}

    public record DeliveryStatusChangedRes(String deliveryId, DeliveryStatus status) {}

    @ZLinkPacket("FindCustomerActorReq")
    public record FindCustomerActorReq(String customerId) {}

    public record FindCustomerActorRes(String customerId, ActorRefSnapshot actorRef) {}

    @ZLinkPacket("EnsureCustomerActorReq")
    public record EnsureCustomerActorReq(String customerId) {}

    public record EnsureCustomerActorRes(String customerId, ActorRefSnapshot actorRef) {}

    @ZLinkPacket("ServerAssertionReq")
    public record ServerAssertionReq(String successfulDeliveryId, String reassignedDeliveryId) {}

    public record ServerAssertionRes(boolean passed, String[] evidence) {}
}

File: samples/kotlin/DeliveryDispatch/Shared/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/shared/contracts/Messages.kt
package systems.zlink.samples.kotlin.deliverydispatch.shared.contracts

import java.time.Instant
import systems.zlink.framework.actors.ActorRefSnapshot

data class CreateDeliveryReq(
    val deliveryId: String,
    val customerId: String,
    val pickupAddress: String,
    val dropoffAddress: String,
)

data class CreateDeliveryRes(val deliveryId: String)

data class SubscribeDeliveryReq(val deliveryId: String)

data class SubscribeDeliveryRes(val deliveryId: String)

data class DeliveryStatusNotify(
    val deliveryId: String,
    val status: DeliveryStatus,
    val courierId: String,
    val occurredAt: Instant,
)

data class AssignDeliveryMsg(
    val deliveryId: String,
    val customerId: String,
    val pickupAddress: String,
    val dropoffAddress: String,
)

data class BindCourierSessionReq(
    val courierId: String,
    val actor: ActorRefSnapshot? = null,
    val sessionRoute: String? = null,
)

data class BindCourierSessionRes(
    val courierId: String,
    val actor: ActorRefSnapshot,
    val sessionRoute: String,
)

data class BindCourierReq(val courierId: String, val sessionRoute: String)

data class BindCourierRes(
    val courierId: String,
    val actor: ActorRefSnapshot,
    val sessionRoute: String,
)

data class FindCourierActorReq(val courierId: String)

data class FindCourierActorRes(val courierId: String, val actor: ActorRefSnapshot?)

data class EnsureCourierActorReq(val courierId: String)

data class EnsureCourierActorRes(val courierId: String, val actor: ActorRefSnapshot)

/**
 * The offer, and the courier's answer to it. Both are one-way (common sample spec section 7.4): a
 * courier looks at a screen and presses a button, and tying that time to a request's reply would
 * hold the spot's serial queue open for as long as the courier thinks. The decision was always
 * going to arrive as a separate inbound message, because the server can only push to a client,
 * never request of it. [attempt] is what makes the pairing safe: a decision that names another
 * attempt arrived too late and is dropped.
 */
data class OfferDeliveryMsg(
    val courierId: String,
    val deliveryId: String,
    val attempt: Int,
    val pickupAddress: String,
    val dropoffAddress: String,
)

data class OfferDeliveryResultMsg(
    val deliveryId: String,
    val courierId: String,
    val attempt: Int,
    val accepted: Boolean,
    val reason: String? = null,
)

data class OfferDeliveryNotify(
    val courierId: String,
    val deliveryId: String,
    val pickupAddress: String,
    val dropoffAddress: String,
)

data class CourierDecisionMsg(
    val deliveryId: String,
    val courierId: String,
    val accepted: Boolean,
    val reason: String? = null,
)

data class DeliveryStatusChangedReq(
    val deliveryId: String,
    val customerId: String,
    val status: DeliveryStatus,
    val courierId: String,
    val occurredAt: Instant,
)

data class DeliveryStatusUpdatedMsg(
    val deliveryId: String,
    val customerId: String,
    val status: DeliveryStatus,
    val courierId: String,
    val occurredAt: Instant,
)

data class DeliveryStatusChangedRes(val deliveryId: String, val status: DeliveryStatus)

data class FindCustomerActorReq(val customerId: String)

data class FindCustomerActorRes(val customerId: String, val actorRef: ActorRefSnapshot?)

data class ServerAssertionReq(val successfulDeliveryId: String, val reassignedDeliveryId: String)

data class ServerAssertionRes(val passed: Boolean, val evidence: Array<String>)

data class EnsureCustomerActorReq(val customerId: String)

data class EnsureCustomerActorRes(val customerId: String, val actorRef: ActorRefSnapshot)

enum class DeliveryStatus {
    Created,
    Assigned,
    Reassigned,
    Accepted,
    PickedUp,
    Delivered,
    Failed,
}
